crypto/merkletrie/cache.go

```diff
@@ -368,60 +516,135 @@ func (mtc *merkleTrieCache) commit() error {
-func (mtc *merkleTrieCache) calculatePageHashes(page int64) (err error) {
+func (mtc *merkleTrieCache) calculatePageHashes(page int64, newPage bool) (fanoutRelocatedNodes int64, err error) {
 	nodes := mtc.pageToNIDsPtr[uint64(page)]
-	for i := page * mtc.nodesPerPage; i < (page+1)*mtc.nodesPerPage; i++ {
-		if mtc.pendingCreatedNID[storedNodeIdentifier(i)] == false {
+	for i := storedNodeIdentifier(page * mtc.nodesPerPage); i < storedNodeIdentifier((page+1)*mtc.nodesPerPage); i++ {
+		if !newPage && mtc.pendingCreatedNID[i] == false {
+			continue
+		}
+		node := nodes[i]
+		if node == nil {
 			continue
 		}
-		node := nodes[storedNodeIdentifier(i)]
-		if node != nil {
-			if err = node.calculateHash(mtc); err != nil {
-				return
+
+		if err = node.calculateHash(mtc); err != nil {
+			return
+		}
+
+		nodeChildCount := node.getChildCount()
+		if nodeChildCount > mtc.maxChildrenPagesThreshold {
+			nodeUniqueChildPages := node.getUniqueChildPageCount(mtc.nodesPerPage)
+			if nodeUniqueChildPages > mtc.maxChildrenPagesThreshold {
+				// see if we can fit all the child nodes into the existing page or not. If not, we might want to start
+				// a new page as long as there is a chance that all the children would be able to fit into that page.
+				if nodeChildCount < uint64(mtc.nodesPerPage) &&
+					mtc.getPageFillFactor(uint64(mtc.mt.nextNodeID)/uint64(mtc.nodesPerPage)) > mtc.targetPageFillFactor {
+					// adjust the next node id to align with the next page.
+					mtc.mt.nextNodeID = storedNodeIdentifier((1 + uint64(mtc.mt.nextNodeID)/uint64(mtc.nodesPerPage)) * uint64(mtc.nodesPerPage))
+				}
+				node.reallocateChildren(mtc)
+				fanoutRelocatedNodes++
 			}
 		}
 	}
 	return
 }
 
+// getPageFillFactor calculates the fill factor for a given page, or return 0 if the page is not in memory.
+func (mtc *merkleTrieCache) getPageFillFactor(page uint64) float32 {
+	if pageMap := mtc.pageToNIDsPtr[page]; pageMap != nil {
+		return float32(len(pageMap)) / float32(mtc.nodesPerPage)
+	}
+	return 0.0
+}
+
+// reallocatePage reallocates an entire page into the latest page(s). It also update the reallocationMap for all the nodes that have been moved,
+// so that we could update the needed node dependencies.
+func (mtc *merkleTrieCache) reallocatePage(page uint64, reallocationMap map[storedNodeIdentifier]storedNodeIdentifier) (reallocatedNodes int) {
+	nextID := mtc.mt.nextNodeID
+	reallocatedNodes = len(mtc.pageToNIDsPtr[page])
+	nextPage := uint64(nextID) / uint64(mtc.nodesPerPage)
+	if reallocatedNodes == 0 {
+		// if we aren't going to reallocate any nodes, no need to allocate (maybe)
+		// new pages for these.
+		goto skipContentDeletion
+	}
+
+	if _, has := mtc.pageToNIDsPtr[nextPage]; !has {
+		pageMap := make(map[storedNodeIdentifier]*node, mtc.nodesPerPage)
+		mtc.reallocatedPages[nextPage] = pageMap
+		mtc.pageToNIDsPtr[nextPage] = pageMap
+		mtc.pagesPrioritizationMap[nextPage] = mtc.pagesPrioritizationList.PushFront(nextPage)
```

Comment:
this can be a local func parameterized with nextPage or lastPage

Reply:
true, but I don't like the idea of making another function here. I think that a better implementation would be to set up a local variable with `nextPage`/`lastPage` and perform the above if that was set.